src/restricted-endpoints/replication/bulk-document/bulk-document.service.ts

```diff
@@ -96,16 +105,79 @@ export class BulkDocumentService {
     );
     return {
       new_edits: request.new_edits,
-      docs: request.docs.filter((doc) =>
-        this.hasPermissionsForDoc(
-          doc,
-          response.rows.find((responseDoc) => responseDoc.id === doc._id),
-          ability,
+      docs: request.docs
+        .filter((doc) =>
+          this.hasPermissionsForDoc(
+            doc,
+            response.rows.find((responseDoc) => responseDoc.id === doc._id),
+            ability,
+          ),
+        )
+        .map((doc) =>
+          this.removeFieldsWithoutPermissions(
+            doc,
+            response.rows.find((responseDoc) => responseDoc.id === doc._id),
+            ability,
+          ),
         ),
-      ),
     };
   }
 
+  private deleteEmptyValues(
+    updatedDoc: DatabaseDocument,
+    existingDoc: DocMetaInf,
+  ) {
+    const fieldKeys = this.getCustomFieldKeys(updatedDoc);
+
+    for (let i = 0; i < fieldKeys.length; i++) {
+      if (
+        updatedDoc[fieldKeys[i]] === '' ||
+        updatedDoc[fieldKeys[i]] === undefined ||
+        updatedDoc[fieldKeys[i]] === null
+      ) {
+        delete existingDoc.doc[fieldKeys[i]];
+        delete updatedDoc[fieldKeys[i]];
+      }
+    }
+
+    return Object.assign(existingDoc ? existingDoc.doc : {}, updatedDoc);
+  }
+
+  private removeFieldsWithoutPermissions(
+    updatedDoc: DatabaseDocument,
+    existingDoc: DocMetaInf,
+    ability: DocumentAbility,
+  ): DatabaseDocument {
+    let action: any;
+
+    if (existingDoc) {
+      if (updatedDoc._deleted) {
+        existingDoc.doc._deleted = true;
+        return existingDoc.doc;
+      } else {
+        action = 'update';
+      }
+    } else {
+      action = 'create';
+    }
+
+    const fieldKeys = this.getCustomFieldKeys(updatedDoc);
+
+    for (let i = 0; i < fieldKeys.length; i++) {
+      if (!ability.can(action, updatedDoc, fieldKeys[i])) {
+        delete updatedDoc[fieldKeys[i]];
+      }
+    }
```

Comment:
Why is this function in a Controller? Did not saw it there.

Reply:
I think I showed it to you initially when we talked about the requirements for this feature. It was only in the controller because we did not yet use it in other places.